lib/ActivityStreams.js

```diff
@@ -290,47 +281,69 @@ ActivityStreams.prototype = {
    */
   _asyncBuildPlacesCache: Task.async(function*() {
     if (simplePrefs.prefs["query.cache"]) {
-      yield Promise.all([
-          this._memoized.getTopFrecentSites(),
-          this._memoized.getRecentBookmarks(),
-          this._memoized.getRecentLinks(),
-          this._memoized.getFrecentLinks(),
-          this._memoized.getHistorySize(),
-          this._memoized.getBookmarksSize(),
-      ]);
-      Services.obs.notifyObservers(null, "activity-streams-places-cache-complete", null);
+      if (!this._populatingCache.places) {
+        this._populatingCache.places = true;
+        yield Promise.all([
+            this._memoized.getTopFrecentSites(),
+            this._memoized.getRecentBookmarks(),
+            this._memoized.getRecentLinks(),
+            this._memoized.getFrecentLinks(),
+            this._memoized.getHistorySize(),
+            this._memoized.getBookmarksSize(),
+        ]);
+        this._populatingCache.places = false;
+        Services.obs.notifyObservers(null, "activity-streams-places-cache-complete", null);
+      }
     }
   }),
 
   /**
    * Builds a preview cache with links from a normal content page load
    */
-  _asyncBuildPreviewCache: Task.async(function*(primeCacheTimeout) {
-    let linksToSend = [];
-    let promises = [];
-
-    promises.push(PlacesProvider.links.getTopFrecentSites().then(links => {
-      linksToSend.push(...links);
-    }));
-
-    promises.push(PlacesProvider.links.getRecentBookmarks().then(links => {
-      linksToSend.push(...links);
-    }));
-
-    promises.push(PlacesProvider.links.getRecentLinks().then(links => {
-      linksToSend.push(...links);
-    }));
-
-    promises.push(PlacesProvider.links.getFrecentLinks().then(links => {
-      linksToSend.push(...links);
-    }));
-
-    yield Promise.all(promises);
-    this._previewProvider.saveNewLinks(linksToSend);
-    this._setBuildPreviewCacheTimeout(primeCacheTimeout);
+  _asyncBuildPreviewCache: Task.async(function*() {
+    if (!this._populatingCache.preview) {
+      this._populatingCache.preview = true;
+      let linksToSend = [];
+      let promises = [];
+
+      promises.push(PlacesProvider.links.getTopFrecentSites().then(links => {
+        linksToSend.push(...links);
+      }));
+
+      promises.push(PlacesProvider.links.getRecentBookmarks().then(links => {
+        linksToSend.push(...links);
+      }));
+
+      promises.push(PlacesProvider.links.getRecentLinks().then(links => {
+        linksToSend.push(...links);
+      }));
+
+      promises.push(PlacesProvider.links.getFrecentLinks().then(links => {
+        linksToSend.push(...links);
+      }));
+
+      yield Promise.all(promises);
+      yield this._previewProvider.asyncSaveNewLinks(linksToSend);
+      this._populatingCache.preview = false;
+      Services.obs.notifyObservers(null, "activity-streams-previews-cache-complete", null);
+    }
   }),
 
   /**
+   * Set up preview cache to be repopulated every 6 hours
```

Comment:
yeah. I would expect that to be cleared out every time the cache gets built. Then this function called again to reset it. Also, as it is right now, it gets set up once and never gets set again after being triggered.